src/tools/extract-conditionals.cpp

```diff
@@ -0,0 +1,991 @@
+            {
+                for (auto it = nodes.first; it != nodes.second; ++it)
+                {
+                    if (std::get<1>(*it) == restriction.to)
+                        continue;
+
+                    callback(location,
+                             std::get<2>(*from),
+                             restriction.via,
+                             std::get<2>(*it),
+                             restriction.tag,
+                             restriction.value,
+                             restriction.condition);
+                }
+            }
+            else
+            {
+                auto to = find_node(nodes, restriction.to);
+                if (to != nodes.second)
+                {
+                    callback(location,
+                             std::get<2>(*from),
+                             restriction.via,
+                             std::get<2>(*to),
+                             restriction.tag,
+                             restriction.value,
+                             restriction.condition);
+                }
+            }
+        }
+    }
+
+  private:
+    using index_type =
+        osmium::index::map::SparseMemArray<osmium::unsigned_object_id_type, osmium::Location>;
+    using adjacency_type =
+        std::tuple<osmium::object_id_type, osmium::object_id_type, osmium::object_id_type>;
+
+    const std::vector<ConditionalRestriction> &restrictions;
+    std::unordered_set<osmium::object_id_type> related_vias;
+    std::vector<adjacency_type> via_adjacency;
+    index_type location_storage;
+};
+
+int RestrictionsDumpCommand(const char *executable, const std::vector<std::string> &arguments)
+{
+    std::string osm_filename, csv_filename;
+    ParseDumpCommandArguments(executable, "cond-dump", arguments, osm_filename, csv_filename);
+
+    // Read OSM input file
+    const osmium::io::File input_file(osm_filename);
+
+    // Read relations
+    std::vector<ConditionalRestriction> conditional_restrictions_prior;
+    ConditionalRestrictionsCollector restrictions_collector(conditional_restrictions_prior);
+    osmium::io::Reader reader1(
+        input_file, osmium::io::read_meta::no, osmium::osm_entity_bits::relation);
+    osmium::apply(reader1, restrictions_collector);
+    reader1.close();
+
+    // Handle nodes and ways in relations
+    ConditionalRestrictionsHandler restrictions_handler(conditional_restrictions_prior);
+    osmium::io::Reader reader2(input_file,
+                               osmium::io::read_meta::no,
+                               osmium::osm_entity_bits::node | osmium::osm_entity_bits::way);
+    osmium::apply(reader2, restrictions_handler);
+    reader2.close();
+
+    // Prepare output stream
+    std::streambuf *buf = std::cout.rdbuf();
+    std::ofstream of;
+    if (!csv_filename.empty())
+    {
+        of.open(csv_filename, std::ios::binary);
+        buf = of.rdbuf();
+    }
+    std::ostream stream(buf);
```

Comment:
It is only to have possibility for piping text streams. I would like keep it, but don't know how to make it simpler without macros